### ISSUE TITLE
Python compilation fails without --pycffi-dynamic-lib-list

### ISSUE THREAD

**Issue by steweg** [dropbox/djinni#237]
Hi all,

we recently tested python branch as we have reached a point where we would like to use python together with our existing C++. So what we did was followed the example in order to try to build required headers. What we found was that first at all instructions are completely missing. Instead we reverse engineered the run_py_example.sh file. We filled whatever we knew and left blanks whatever we have no idea.

The result was that trying to run line 22 of given script gives us an error like this:
```
x86_64-linux-gnu-gcc: error: ./myTestApp/wrappers_cffi.cpython-35m-x86_64-linux-gnu.so: No such file or directory
Traceback (most recent call last):
  File "/usr/lib/python3.5/distutils/unixccompiler.py", line 207, in link
    self.spawn(linker + ld_args)
  File "/usr/lib/python3.5/distutils/ccompiler.py", line 909, in spawn
    spawn(cmd, dry_run=self.dry_run)
  File "/usr/lib/python3.5/distutils/spawn.py", line 36, in spawn
    _spawn_posix(cmd, search_path, dry_run=dry_run)
  File "/usr/lib/python3.5/distutils/spawn.py", line 159, in _spawn_posix
    % (cmd, exit_status))
distutils.errors.DistutilsExecError: command x86_64-linux-gnu-gcc failed with exit status 1
```
After a while of troubleshooting we find out that the problem is line in djinni generated file pycffi_lib_build.py:
```
ffi.set_source('myTestApp.wrappers_cffi', '''#include "clean_headers.h"''',
    runtime_library_dirs=['.'],
    libraries=[''])
```
This libraries line was generated because we left  --pycffi-dynamic-lib-list option empty, which lead in the end to create linker command for gcc with the following attributes:
`-l -o  ./myTestApp/wrappers_cffi.cpython-35m-x86_64-linux-gnu.so`
Notice the "-l" without any file. GCC/ld treats very problematic. Therefore what we did was 
replace the auto generated part with just:
```
ffi.set_source('myTestApp.wrappers_cffi', '''#include "clean_headers.h"''',
    runtime_library_dirs=['.'])
```

And the compilation worked like charm

So what needs to be done is either making option --pycffi-dynamic-lib-list mandatory for python part (please add comments what should go there if needed) or adjust the djinni code to avoid empty "libraries" line of code

And one more critical part is missing as well. As we have followed the example, we find out that path to "clean_headers.h" must be provided prior trying to compile (we adjusted our CPATH environment variable to allow GCC to compile properly). Again this is never mentioned anywhere and to pin-point the root-cause, one can easily spend a lot of man-hours.

**Comment by steweg** [comment]
We just found one more thing. There is missing header include for std::unique_ptr. This patch [include.patch](https://github.com/dropbox/djinni/files/318458/include.patch.txt) fixes it.

**Comment by artwyman** [comment]
I'm excited to hear you're trying out the Python support.  Sorry it's been tricky for you.  That language support has suffered somewhat from not having a dedicated maintainer, since we got it to PoC/Beta phase and then put it aside.  We don't have an internal use case for it at Dropbox, and are likely not to try to tackle further work on it until we do have one.  That being said, I welcome PRs to fix the two issues you identified, both of which sound pretty simple to fix.  I can help out with diagnosing and fixing any other problems you run into, but I'm not likely to have significant time to dedicate to development on Python in the near future.

Regarding documentation, I assume you found [README.Python.md](https://github.com/dropbox/djinni/blob/python/README.Python.md), but you're right that it doesn't go as in-depth as maybe it should.  If you want to expand it with what you've learned, I'm happy to review and merge PRs on that.

**Comment by artwyman** [comment]
I believe this is fixed by #240
